deblobbify, replacing the system folder with several other components

robertkirkman opened the pull request:
- Fixes #35
- Fixes #60 
- Fixes #63 
- Fixes #65
- Fixes #68

Summary of changes:

- Deletes the entire `system` folder and replaces it with a script `generate.sh` that can regenerate the entire `system` folder.
- The root filesystem, including `system` folder, is regenerated using purely a Termux bootstrap `.zip`, and `.deb` files that can be built using the `termux-packages` build and CD system, then retrieved from the `termux.dev` package host.
- Adds an environment variable `$TERM` that is set to a value of "`xterm`" for the `clear` command to have convenient compatibility with most terminal emulators, and this has also been tested with Kitty and TTY.
- Simplifies the instances of the string "`com.termux`" into a single instance for slightly more convenient use with custom bootstraps.
- Fixes the existing `docker.io/termux/termux-docker:aarch64` image _not_ working on `aarch64` devices that happen to have an esoteric **"[Cortex-A34](https://developer.arm.com/Processors/Cortex-A34)"-type or similar processor** due to the elimination of [this 32-bit ARM binary blob](https://github.com/termux/termux-docker/blob/e3c861f5b6cb762d3d99f6d6e6dd78cdffe109b6/system/arm/bin/busybox), which Cortex-A34-like processors (including those in Apple Silicon devices) do not have hardware support for executing, unlike "[Cortex-A72](https://developer.arm.com/Processors/Cortex-A72)"-like processors.
- Replaces Busybox with a real build of Toybox based on the source code of AOSP 9.0.0 that has been patched to meet the requirements of termux-docker, including providing `su` to log in as the `system` user within `entrypoint.sh`.
- Uses `mksh` and `bash` alone to implement `sh`, instead of Busybox's `sh`. 
- Uses a build of `iputils` based on the source code of AOSP 9.0.0 to implement `/system/bin/ping` and `/system/bin/ping6`.
- Uses the preexisting `termux-packages` `root/dnsmasq` to implement `dnsmasq`, which works after some adjustments to its invocations.
- Adds `libandroid-stub`, which is described in multiple places as intended specifically for use with termux-docker.
  - I have confirmed that it indeed makes `ffmpeg` work inside termux-docker, fixing this error: `linker: CANNOT LINK EXECUTABLE "ffmpeg": library "libandroid.so" not found`
  - [in its `build.sh`](https://github.com/termux/termux-packages/blob/2499f601b33ce3c5fe9458ce8ffc45b73d874502/packages/libandroid-stub/build.sh#L39-L45)
  - [xtkoba believed that `libandroid-stub` should be preinstalled in termux-docker](https://github.com/termux/termux-packages/issues/16902#issuecomment-1592234036)
- Deletes `/system/etc/ld.config.28.txt` without replacing it with anything, since it does not appear to be necessary for anything in any test so far, but it could be easily re-added if it turns out that there is something that needs it.
  - I think that `ld.config.28.txt` is probably only necessary on ROMs that have a `/vendor` folder and stuff inside (["VNDK"](https://source.android.com/docs/core/architecture/vndk/snapshot-design#vndk-snapshot-arch)), termux-docker does not have a `/vendor` folder or support anything related to that so, for now it seems unnecessary
- Adds a third method of working around the `personality()` system call permissions issue, a custom build of Docker, which is not required on all ARM devices, but which might be required sometimes, and is required on my `armv7l` device that the `--privileged` and `--security-opt` arguments both failed to work on.
- Replaces the use of QEMU in GitHub Actions with the new official public GitHub ARM runner.
- Enables updating the termux packages within the container before publishing the ARM images.
- Changes `latest` tag to `x86_64` instead of `i686`.
  - requested by Ted and Biswa96
- Refactors `entrypoint_root.sh` and `entrypoint.sh` to have more similar guard clause structure to each other, for readability.
- Adds termux-pacman support
- Removes symbolic link from `/usr` to `$PREFIX` because real Android devices running Termux App do not have that symbolic link, so removing this makes termux-docker more accurate to the Termux App build environment, and also would engage the same `termux-exec` codepath that real Termux App uses to intercept only specific shebangs. More information: https://github.com/termux/termux-packages/pull/25995#discussion_r2328774669
- Increases the bootstrap version from 2023.02.19 to 2026.04.05


> [!IMPORTANT]
> In some (not all) cases, the **custom build of Docker with modified `personality()` system call permissions** is required. It is unknown how to predict whether a device will require that or not, but **it is suspected it might have something to do with the host kernel**, which is a factor whose properties it is understood tend to vary dramatically between different ARM devices. **This affects termux-docker both before and after this PR, so this PR does not introduce the problem, but it provides a `.patch` file to help mitigate it**.


Tested both locally and in GitHub Actions:

- `aarch64`: working
- `x86_64`: working
- `arm`: working*
- `i686`: working

> *both before and after this PR, `arm` architecture containers have been observed to be affected by the `personality()`-related issue and require the custom build of Docker more often than the other architectures, but not always, and are not the only architecture that can be affected.

~~The "temporary" things like the `temporary-package-storage` folder and the blobs inside are only necessary for testing while this is not merged~~ **done**:

- https://github.com/termux/termux-packages/pull/22906

The large amount of changes to the internal components of termux-docker could introduce unpredictable bugs or unexpected behavior that is not present with the current official termux-docker. For that reason, I hope to learn **all of the current intended use cases that termux-docker is officially supposed to be used for**, so that I can test my implementation with them and add additional software or fixes if needed.

Known Unique Limitations:

- For some reason, the `-c` argument of `toybox`-`su` is currently non-functional. For the current featureset of termux-docker, both before and after this PR, that does not seem to be a major priority because termux-docker currently _only_ supports the `root` user via the `entrypoint_root.sh` and _not_ via `su` or `tsu` from within the default `entrypoint.sh`, i.e. most `termux-packages` `root-packages` cannot currently be conveniently tested using termux-docker anyway. If "proper simulation of a rooted device", so to speak, is desirable in the future, to facilitate more convenient testing of `root-packages`, if ever necessary, that could be investigated and handled by a different issue+PR at a later date.

twaik commented:
> modifying the termux-packages `bionic-host` package is not acceptable

Nobody said it is not acceptable. I think nobody tried it. 
I created the package as a solution for running android executables during regular building with toolchain to avoid using hostbuild step, but for some reason qemu does not want to run fine with aarch64 executables on amd64 host, at least in some cases.
Also it fits fine for using in `termux-docker` but it was not modified to use `bionic-host` package.

robertkirkman commented:
> for some reason qemu does not want to run fine with aarch64 executables on amd64 host, at least in some cases

I am not sure exactly what exact scenario you are talking about here, or what error it is you were seeing at the time, if you mean a scenario with a `build.sh` involving only the regular cross-compiling toolchain, then what I am about to say is probably not relevant,

But just in case knowing this helps you, there is a way to emulate an ARM processor on an x86 device that is **very high in emulation accuracy and behaves almost identically to how real ARM devices behave**.

When carefully configured, I have never seen `qemu-system-aarch64` fail to execute ARM software on an x86 processor, as long as that same ARM software also works on real ARM devices. If I ever see frustrating errors while trying to use **user-mode emulation** of any kind, `qemu-system-aarch64` is my go-to solution for accuracy to the exact behavior of real ARM devices, when I am in a place with only x86 devices and accessing real ARM devices is inconvenient.

Since there are very few good guides on the subject, and almost all other guides skip mentioning lots of necessary steps, leaving everything unexplained and causing most people to feel like `qemu-system-aarch64` is much too tedious to consider using on a regular basis, I created [this guide](https://gist.github.com/robertkirkman/f79441c79811ad263f2f881f7864e793) that covers every detail of the setup process in a very straightforward, completely explained, and future-proofed way based on the latest UEFI technology, which is also distro-agnostic to both the host and the guest, making it very clear what changes when the guest and host are mixed and matched.

twaik commented:
I will just link this.
https://github.com/termux/termux-packages/issues/18517#issuecomment-1814322939
I know it is not a simple scenario, but GIR-generated binary used for generating some XMLs for some reason hangs and CI is terminated with timeout after 6 hours so it is a no go for us. It was the very first and the last attempt (my attempt) to use cross-compiled binaries for building (to avoid using `host_build` step).

robertkirkman commented:
>  GIR-generated binary used for generating some XMLs for some reason hangs and CI is terminated with timeout after 6 hours so it is a no go for us

I experimented with some local tests based on `qemu-system` technology  (as opposed to `qemu-user` technology), 

and while it can inherently avoid several of the issues described in that thread, such as the GIR hanging problem, the problems with any other binaries hanging, and the `pid_max` problem, due to the higher emulation accuracy and environment micromanagement of `qemu-system` compared to `qemu-user`, 

**very unfortunately, it cannot overcome the slowness problem that licy183 explained. Additionally, the higher emulation accuracy of `qemu-system` comes with a cost that the slowness is even more extreme, causing the build time for multiple things I need to build to go far over the 6 hours limit** (my builds were successful locally, but took 16 hours). Even though that technically works for me locally, that makes it not viable for anything in GitHub Actions without also using the 6 hour limit bypass incremental build workaround technique every time.

I think the only efficient possible future plan on this topic, is to make use of termux-docker, or licy183's tur-avd, or another minimal android container, combined both with the new GitHub ARM runners  and with the termux-packages `clang` package, to build some packages with cross compilation disabled and emulation also disabled.

I am already testing the GitHub ARM runner in this PR and it works for example here, 
but since TUR has already been building packages in this same codepath using self-hosted ARM runners instead of cloud ARM runners, it is possible that maybe licy183 is already about to or in the process of adding the cloud ARM runners to TUR, which could serve as an example of how to proceed.
[They had tested a PR using it a few months ago](https://github.com/termux-user-repository/tur/pull/1132), but I did not see any other tests of it yet. 
If they do not end up adding it soon, I could attempt my own way of adding support for it to the termux-packages repo, but first I need to finish the changes to termux-docker.

~~I cannot complete the changes to termux-docker without first removing blockers, like fixing the builds of `pypy` and `pypy3` so that the dependency can be merged. Because of that, I think I am currently forced to first use the only workaround possible to force them to build in a single PR with the updated `aosp-libs` package, the `pid_max` change,  then I could attempt to migrate the ARM builds of `pypy` and `pypy3` to an actual ARM runner workflow some time afterward, after creating it. The `pid_max` change would remain required for running any 32-bit termux-docker-based workflows anyway, so it could be moved to the termux-docker-based workflow instead at the time that is finished.~~ **done**

I have tested locally that `pypy` would build and work using termux-packages `clang` inside a complete termux-docker ARM system, it is just that there is no workflow set up for that yet. I would prefer waiting to try introducing termux-docker to such a workflow until after it has had the blobs removed.

robertkirkman commented:
I have managed to encounter a test case where:

- normal devices are negatively affected by having `libandroid-stub` installed
- under equivalent conditions within termux-docker, `libandroid-stub` instead only helps, and the error from the normal device cannot be reproduced

That seems to support the assumption that the long-preexisting `libandroid-stub` package is probably entirely beneficial inside termux-docker with no anticipated downsides, so it makes sense to preinstall it.

#### How to reproduce the negative effect of `libandroid-stub` on a normal device

1. Use `pkg install libandroid-stub`
2. Use `pkg install mpv`
3. `mpv`

- Result

```
CANNOT LINK EXECUTABLE "mpv": cannot locate symbol "AImageReader_newWithUsage" referenced by "/system/lib64/libhwui.so"...
```

#### How to reproduce the above conditions inside termux-docker and observe no error

1. Use `pkg install libandroid-stub` (if not already installed)
2. Use `pkg install mpv`
3. Work around #70 by installing unofficial package `libopensles-standalone` as explained in that issue
4. `mpv`

- Result

```
mpv 0.39.0 Copyright © 2000-2024 mpv/MPlayer/mplayer2 projects
 built on Feb 26 2025 00:33:35
...
```

twaik commented:
About libandroid-stub. Currently we package `libandroid.so` from NDK, but can we make our own version of libandroid which will dlopen+dlsym regular libandroid.so in the case if it exists and call it's methods and return some error code in the case if symbols are not present or libandroid is not available?

twaik commented:
I mean in this case we will be able to ship `libandroid-stub` regardless of having android or docker and it should work fine on devices. But I am not sure about this.

robertkirkman commented:
I am not sure, because I think I do not have one of the devices that (i assume) has a `/system/lib64/libandroid.so`, but where the `libandroid-stub` package is still needed, so I cannot test on that device.

If I can find a device or contact someone who owns such a device that contains a `/system/lib64/libandroid.so` but which does not work with the `ffmpeg` package unless `libandroid-stub` is installed, then maybe I could try to make code until I make it work on termux-docker, and also that device, and also normal devices all at the same time from one package.

- Here was one of them https://github.com/termux/termux-packages/issues/20334#issuecomment-2135903724

Samsung Galaxy M52

robertkirkman committed:
deblobbify, replacing the system folder with several other components

robertkirkman commented on generate.sh:
```diff
@@ -0,0 +1,307 @@
+done
+
+# read_package_lists and pull_package are based on their implementations
+# in https://github.com/termux/termux-packages/blob/7a95ee9c2d0ee05e370d1cf951d9f75b4aef8677/scripts/generate-bootstraps.sh
```
agnostic-apollo has recently stated that they are working on a large update to `generate-bootstraps.sh` that will probably change the code that this part was copied from, so I will prefer to wait for that to come so I can update this implementation to match the changes that will come to the future new `generate-bootstraps.sh`.

Jzhenli commented:
@robertkirkman seems all checks have passed, so now it's possible to use github action to build software in termux, right?

robertkirkman commented:
> @robertkirkman seems all checks have passed, so now it's possible to use github action to build software in termux, right?

It is already possible to use github actions to build software in termux by following the example made by me and guijan here:

https://github.com/robertkirkman/termux-on-gha

however just keep in mind that by doing so, the image used to build currently contains binaries of unverifiable origin. this PR, after it is merged, will remove binaries of unverifiable origin and replace them with verified ones.

Jzhenli commented:
@robertkirkman thanks, hope this PR can be merged ASAP.

robertkirkman commented:
It could be, but the reason I decided to wait is because, you can see that in this PR there is a lot of code I copied and pasted from this file 

https://github.com/termux/termux-packages/blob/d7c678e5133deeeea1e88c3b0ad1ff72f5915a4f/scripts/generate-bootstraps.sh#L61

and I found out recently that that file is going to get an update that will probably change a  lot of it,

so I didn't want to add something here that would be outdated soon, I wanted to wait for the changes to the `generate-bootstraps.sh` that might appear. That way I could copy and paste all the new changes into this PR and fully test them here too, before this gets merged.

robertkirkman commented:
> I hope to learn all of the current intended use cases that termux-docker is officially supposed to be used for, so that I can test my implementation with them and add additional software or fixes if needed.

After recent build fixes in TUR and termux-packages, This PR is now confirmed to produce a termux-docker that can successfully build all current tur-on-device packages (packages in the tur-on-device folder of TUR), like this:

```bash
git clone https://github.com/termux-user-repository/tur.git
cd tur/
./setup-environment.sh
scripts/setup-termux.sh
./build-package.sh -I -f blender \
                         classic-flang \
                         gnome-chess \
                         gnome-mahjongg \
                         gnome-sudoku \
                         hello-tur-on-device \
                         lpsolve \
                         nqp \
                         perl-net-ssleay \
                         pixz \
                         python2-numpy \
                         python2-scipy \
                         r-base \
                         rakudo \
                         s-nail \
                         thc-hydra
```

The previous build problems all also affected the current official termux-docker, but now that all are building regardless of which kind of termux-docker is used, it is confirmed this PR's version of termux-docker is 100% compatible with tur-on-device, so this PR would not break anything for TUR when merged.

robertkirkman commented:
It has been a while and `generate-bootstraps.sh` has not changed yet, so I have decided that it's not really necessary to wait for it to change before applying this PR, because this PR contains its own copies of the https://packages-cf.termux.dev/apt/termux-main/dists/stable/main and https://packages-cf.termux.dev/apt/termux-root/dists/root/stable URLs, so as long as those URLs remain the same, changes to `generate-bootstraps.sh` will not externally affect this proposed termux-docker version's `generate.sh`.

robertkirkman committed:
implement pacman support

robertkirkman committed:
implement support for building the image on BSD

Some BSD-based operating systems support Linux Docker images by
implementing a `docker` command that has a backend powered by
virtualization and an internal virtualized Linux kernel.

The result is that all code running inside the container is
only exposed to a Linux kernel and is unaware of BSD,
but all code running outside the container is directly exposed
to the BSD-based environment

BSD-based operating systems used to develop this:

- MacOS
- FreeBSD
- OpenBSD

It's often hard to find where exactly the `gtar`, `gar`, and `bash`-version-4+ commands are on BSD-based operating systems,
so if you see errors related to those, try to find and install the dependencies for your OS and then add them to `$PATH`.

robertkirkman committed:
bump `custom-docker-with-unrestricted-personality.patch` to Debian 13

robertkirkman commented:
I am upgrading my use of 32-bit ARM termux-docker from Debian 12 to Debian 13, but in this PR I had already implemented it for Debian 12, which some people might still have, so it seems best not to squash the bump into other commits. This makes it easier for someone to get the Debian 12 version if they need.

robertkirkman committed:
update documentation to mention that x86 containers can, in rare cases, be affected by the `personality`-related error

- I have discovered that it is possible to construct a convoluted situation in which a purely 64-bit x86 container running on a purely 64-bit x86 device can also be affected by the `personality`-related error, and that the `custom-docker-with-unrestricted-personality.patch` is also capable of preventing the error in that rare situation. Contact me if you would like to learn the steps to reproduce that situation.

robertkirkman commented on custom-docker-with-unrestricted-personality.patch:
```diff
@@ -0,0 +1,315 @@
+This removes all restrictions from the personality() system call from within Docker, and
+is only necessary on some specific devices, including some ARM devices but not all ARM devices,
+and including some x86 devices in some rare situations but not all situations,
```
After 11 months of testing, I discovered yesterday that, in a particular convoluted situation, it is possible to reproduce the `personality`-related error while using a purely 64-bit x86 termux-docker container on a purely 64-bit x86 Docker engine on a purely 64-bit x86 hardware device with no emulators running at any point,

and that the `custom-docker-with-unrestricted-personality.patch` helps just as well in that situation as it does in all the other situations it helps with,

so I have added notes to the documentation to reflect this discovery.

The setup to reproduce the error on the x86 device with no emulators is particularly convoluted. If someone wants to learn the reproduction steps, contact me.

robertkirkman replied:
Reproduction steps:

- https://github.com/termux/termux-packages/issues/29690#issuecomment-4426425977

robertkirkman commented:
ok, @truboxl I'm sorry for pinging, I wondered do you think this PR is ok, or is there anything wrong with it?

sylirre reviewed (commented):
@robertkirkman Need to upgrade bootstrap version for apt, because apt build uses version from 2023.

Dnsmasq startup check needs to be modified in entrypoint script because sometimes utility needs more time to start and false error shown. But this is not really related to your changes.

Otherwise seems to be good for merging.

sylirre commented on generate.sh:
```diff
@@ -0,0 +1,469 @@
+case "${TERMUX_PACKAGE_MANAGER}" in
+	apt)
+		TERMUX_DOCKER__IMAGE_NAME="termux/termux-docker"
+		TERMUX_DOCKER__BOOTSTRAP_VERSION="2023.02.19-r1%2Bapt-android-7"
```
This

robertkirkman replied:
Ok sure, I will upgrade this. I agree.

To reveal my reasoning, I wanted to let you know that the reason I did not change the version of the bootstrap is because, a long time ago, someone opened a PR where they tried to increase the version of the bootstrap and it was closed as irrelevant! https://github.com/termux/termux-docker/pull/67#issuecomment-2245965114

obviously that PR was different from this PR, but I was trying to be "on the safe side" , thank you for confirming I can increase the bootstrap version.

sylirre commented on generate.sh:
```diff
@@ -0,0 +1,469 @@
+		;;
+	pacman)
+		TERMUX_DOCKER__IMAGE_NAME="termux/termux-docker-pacman"
+		TERMUX_DOCKER__BOOTSTRAP_VERSION="2026.02.01-r1%2Bpacman-android-7"
```
This also better to keep fresh

robertkirkman commented:
> Dnsmasq startup check needs to be modified in entrypoint script because sometimes utility needs more time to start and false error shown. But this is not really related to your changes.

I agree, that can sometimes happen. However, I wanted to confirm - you tested this version and you still were able to reproduce the problem with dnsmasq taking slightly too long to start, right?

I haven't been able to reproduce that on any of my systems while using the changes in this PR, so it seemed to me like it stopped or greatly reduced in severity. Maybe the severity only reduced, so it still happens to you?

If so, do you think increasing its delay from 1 second to 2 seconds will be long enough?

robertkirkman committed:
bump apt and pacman bootstraps to 2026.04.05

sylirre commented:
Yes, I tested only this pull request.

I think the solution can be different, like looping until dnsmasq pidfile appears. But this will require another fix as image from this pr ships a pidfile from one of dockerfile steps.

Though this doesn't block changes from merging.

robertkirkman commented:
> I think the solution can be different, like looping until dnsmasq pidfile appears. But this will require another fix as image from this pr ships a pidfile from one of dockerfile steps.

I could try that, but, will you be able to help me test my attempts if I try that? Since, unfortunately I might be unable to reproduce the dnsmasq delay problem anymore, and I would need someone who can reproduce it to know whether my solution would help at all.

Thank you very much!

I am just leaving this note here because I wanted to see a record of what size change would occur in the images (for reference to check in the future if it becomes necessary to)

<img width="2364" height="1069" alt="image" src="https://github.com/user-attachments/assets/76e1e239-bcdf-49b8-aecf-f1bd6f9613d7" />

I believe that I actually added a significant amount of functionality and new commands, but for some reason only the aarch64 image significantly increased in size and the x86 images actually decreased in size. That is better than I anticipated.

robertkirkman commented:
Let's try to fix that next,

- https://github.com/termux/termux-docker/pull/80

that works for me, but does that work for you to wait for `dnsmasq`?
